optuna/visualization/matplotlib/_contour.py

```diff
@@ -489,135 +476,34 @@ def _interpolate_zmap(zmap: Dict[complex, float], contour_plot_num: int) -> None
-) -> float:
-
-    max_fractional_delta = 0.0
-
-    for coord in coordinates:
-        # we will iterate the grid in order
-        # established by `find_coordinates_where_empty`
-        # this will ensure that we are not trying to interpolate
-        # value with zero neighbors
-        current_val = zmap.get(coord, None)
-        max_neighbor = -np.inf
-        min_neighbor = np.inf
-        sum_neighbors = 0.0
-        n_neighbors = 0
-
-        for offset in NEIGHBOR_OFFSETS:
-            # if we represent current value as a center
-            # of 3x3 matrix, neighbors are all other
-            # values excluding those at diagonals
-            neighbor = zmap.get(coord + offset, None)
-
-            if neighbor is None:
-                # off the edge or not filled in
-                continue
-
-            sum_neighbors += neighbor
-            n_neighbors += 1
-
-            if current_val is not None:
-                max_neighbor = max(max_neighbor, neighbor)
-                min_neighbor = min(min_neighbor, neighbor)
-
-        # fill value is just mean of its neighbors
-        new_val = sum_neighbors / n_neighbors
-
-        if current_val is None:
-            zmap[coord] = new_val
-            max_fractional_delta = 1.0
-
-        else:
-            zmap[coord] = (1 + overshoot) * new_val - overshoot * current_val
-            if max_neighbor > min_neighbor:
-                fractional_delta = abs(new_val - current_val) / (max_neighbor - min_neighbor)
-                max_fractional_delta = max(overshoot, fractional_delta)
-
-    return max_fractional_delta
+    # Plotly's algorithm is equivalent to solve the following linear simultaneous equation.
+    # It is discretization form of the Poisson equation.
+    #
+    #     z[x, y] = zmap[(x, y)]                                  (if zmap[(x, y)] is given)
+    # 4 * z[x, y] = z[x-1, y] + z[x+1, y] + z[x, y-1] + z[x, y+1] (if zmap[(x, y)] is not given)
+
+    a_data = []
+    a_row = []
+    a_col = []
+    b = np.zeros(contour_plot_num ** 2)
+    for x in range(contour_plot_num):
+        for y in range(contour_plot_num):
+            grid_index = y * contour_plot_num + x
+            if (x, y) in zmap:
+                a_data.append(1)
+                a_row.append(grid_index)
+                a_col.append(grid_index)
+                b[grid_index] = zmap[(x, y)]
+            else:
+                for dx, dy in ((-1, 0), (1, 0), (0, -1), (0, 1)):
+                    if 0 <= x + dx < contour_plot_num and 0 <= y + dy < contour_plot_num:
+                        a_data.append(1)
+                        a_row.append(grid_index)
+                        a_col.append(grid_index)
+                        a_data.append(-1)
+                        a_row.append(grid_index)
+                        a_col.append(grid_index + dy * contour_plot_num + dx)
+
+    z = scipy.sparse.linalg.spsolve(scipy.sparse.csc_matrix((a_data, (a_row, a_col))), b)
```

Comment:
Is there any reason to use the CSC matrix instead of the CSR matrix?

Reply:
I'm not sure which format is better to use. This PR follows `spsolve`'s example and implementation.
https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.linalg.spsolve.html
https://github.com/scipy/scipy/blob/v1.7.1/scipy/sparse/linalg/dsolve/linsolve.py#L139-L145

Reply:
My simple benchmark shows a selection of data format doesn't have a significant impact.
code: https://colab.research.google.com/gist/himkt/947d7f2bb25d54dac8686fef42eb6521/naive-benchmark-csc-and-csr.ipynb
patch: https://github.com/not522/optuna/compare/fast-contour...himkt:oneshot/fast-contour?expand=1